worker/sshserver/listener_test.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshserver_test
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/testing"
+	"google.golang.org/grpc/test/bufconn"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/worker/sshserver"
+)
+
+type listenerSuite struct {
+	testing.IsolationSuite
+}
+
+var _ = gc.Suite(&listenerSuite{})
+
+func (s *listenerSuite) TestAcceptOnceListener(c *gc.C) {
+	listener := bufconn.Listen(8 * 1024)
+
+	acceptOnceListener, closeAllowed := sshserver.NewAcceptOnceListener(listener)
+	c.Assert(acceptOnceListener, gc.NotNil)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		<-closeAllowed
+		listener.Close()
+	}()
+
+	// Artificial sleep to ensure close is definitely waiting on <-closeAllowed
+	time.Sleep(100 * time.Millisecond)
```

Comment:
this does not ensure that we're waiting on closeAllowed..  you might need to rethink this test.